src/ethereum/london/vm/__init__.py

```diff
@@ -82,10 +85,64 @@ class Evm:
     running: bool
     message: Message
     output: Bytes
-    accounts_to_delete: Dict[Address, Address]
+    accounts_to_delete: Set[Address]
+    touched_accounts: Set[Address]
     has_erred: bool
-    children: List["Evm"]
     return_data: Bytes
     error: Optional[Exception]
     accessed_addresses: Set[Address]
     accessed_storage_keys: Set[Tuple[Address, Bytes32]]
+
+
+def incorporate_child_evm_successful(evm: Evm, child_evm: Evm) -> None:
+    """
+    Incorporate the state of a successful `child_evm` into the parent `evm`.
+
+    Parameters
+    ----------
+    evm :
+        The parent `EVM`.
+    child_evm :
+        The child evm to incorporate.
+    """
+    evm.gas_left += child_evm.gas_left
+    evm.logs += child_evm.logs
+    evm.refund_counter += child_evm.refund_counter
+    evm.accounts_to_delete.update(child_evm.accounts_to_delete)
+    evm.touched_accounts.update(child_evm.touched_accounts)
+    if account_exists_and_is_empty(
+        evm.env.state, child_evm.message.current_target
+    ):
+        evm.touched_accounts.add(child_evm.message.current_target)
+    evm.accessed_addresses.update(child_evm.accessed_addresses)
+    evm.accessed_storage_keys.update(child_evm.accessed_storage_keys)
+
+
+def incorporate_child_evm_error(evm: Evm, child_evm: Evm) -> None:
+    """
+    Incorporate the state of an unsuccessful `child_evm` into the parent `evm`.
+
+    Parameters
+    ----------
+    evm :
+        The parent `EVM`.
+    child_evm :
+        The child evm to incorporate.
+    """
+    # In block 2675119, the empty account at 0x2 (the RIPEMD160 precompile) was
+    # cleared despite running out of gas. This is an obscure edge case that can
+    # only happen to a precompile.
+    # According to the general rules governing clearing of empty accounts, the
+    # touch should have been reverted. Due to client bugs, this event went
+    # unnoticed and 0x2 has been exempted from the rule that touches are
+    # reverted in order to preserve this historical behaviour.
+    if RIPEMD160_ADDRESS in child_evm.touched_accounts:
+        evm.touched_accounts.add(RIPEMD160_ADDRESS)
+    if (
+        child_evm.message.current_target == RIPEMD160_ADDRESS
+        and account_exists_and_is_empty(
+            evm.env.state, child_evm.message.current_target
+        )
+    ):
+        evm.touched_accounts.add(RIPEMD160_ADDRESS)
```

Comment:
To avoid the parentheses hell:

```suggestion
    if child_evm.message.current_target == RIPEMD160_ADDRESS:
        if account_exists_and_is_empty(evm.env.state, child_evm.message.current_target):
            evm.touched_accounts.add(RIPEMD160_ADDRESS)
```